### PR TITLE
ARROW-14183: [C++] Improve select_k_unstable performance

### PR DESCRIPTION
**Pull request by AlvinJ15**
Improve select_k_unstable performance, for tables using radix sort for sort individual batches, and a merge sort with K first elements.

### CODE REVIEW

**Comment by AlvinJ15** [comment]
@ursabot please benchmark

**Comment by cyb70289** [comment]
Looks this PR is still WIP?
Do you have benchmark data on your local machine? Looks there's no obvious change from conbench result. Maybe we should add more benchmarks first?

**Comment by AlvinJ15** [comment]
Yes @cyb70289 I have different runs on my local.
**master branch with** `K=10`
image
**RadixBatch + mergeK**
image

**master branch with** `K=20`
image
**RadixBatch + mergeK**
image

**master branch with** `K=1000`
image
**RadixBatch + mergeK**
image

**master branch with** `K=10000`
image
**RadixBatch + mergeK**
image

**master branch with** `K=N/8`
image
**RadixBatch + mergeK**
image

**Comment by pitrou** [comment]
Should you close the older PR for the same issue?

File: cpp/src/arrow/compute/kernels/chunked_internal.h [hunk 1]
```diff
@@ -144,6 +144,19 @@ struct ChunkedArrayResolver : protected ChunkResolver {
         checked_cast<const ArrayType*>(chunks_[loc.chunk_index]), loc.index_in_chunk);
   }
 
+  template <typename ArrayType>
+  ResolvedChunk<ArrayType> Resolve(ChunkLocation loc) const {
+    return ResolvedChunk<ArrayType>(
+        checked_cast<const ArrayType*>(chunks_[loc.chunk_index]), loc.index_in_chunk);
+  }
+
+  template <typename ArrayType>
```
Comment (aocsa) [inline cpp/src/arrow/compute/kernels/chunked_internal.h:153]: I would suggest to add some documentation to these internal but exposes functions (Resolve, ResolveChunkLocation).

File: cpp/src/arrow/compute/kernels/vector_sort.cc [hunk 2]
```diff
@@ -1734,105 +1768,155 @@ class TableSelecter : public TypeVisitor {
...
+      const auto& batch = *batches[i];
+      end_offset += batch.num_rows();
+      RadixRecordBatchSorter sorter(indices_begin + begin_offset,
+                                    indices_begin + end_offset, batch, sort_options);
+      ARROW_ASSIGN_OR_RAISE(sorted[i], sorter.Sort(begin_offset));
+      DCHECK_EQ(sorted[i].overall_begin(), indices_begin + begin_offset);
+      DCHECK_EQ(sorted[i].overall_end(), indices_begin + end_offset);
+      DCHECK_EQ(sorted[i].non_null_count() + sorted[i].null_count(), batch.num_rows());
+      begin_offset = end_offset;
+      // XXX this is an upper bound on the true null count
```
Comment (aocsa) [inline cpp/src/arrow/compute/kernels/vector_sort.cc:1814]: Remove XXX

**Review by aocsa — CHANGES_REQUESTED** [review]
This PR looks an interesting change in favor of better performance. I put some general comments, however just for clarity I would suggest to show better the benchmark results as well the improvement in numbers (%)  and put some comments in which cases you are getting better/worse results. Moreover if these cases are localized I would suggest also to add new unit tests with comments around this.

**Comment by cyb70289** [comment]
@AlvinJ15 , are you still working on this PR?

**Comment by AlvinJ15** [comment]
> @AlvinJ15 , are you still working on this PR?

@cyb70289 I will rebase and finish this PR for this week, thanks for the remainder

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

**Pull request closed without merging**